Add configuration of authentication provider

lfdebrux opened the pull request:
We want to be able to use other authentication strategies in forms-admin. This PR refactors the application controller to allow configuration of the strategy using the settings files or environment variable.

We also deprecate the old way of selecting the basic authentication strategy, and move the basic authentication strategy out the application controller and into a Warden strategy.

Note that for the setting, we call it `auth_provider`, not `auth_strategy`. The term strategy is used consistently in Warden, but in OmniAuth provider and strategy are used somewhat interchangeably. I think arguably you could say the provider is the external identity provider and the strategy is the code you use to communicate with it, but in the context of a configuration setting there isn't too much difference in it. To be consistent with what we save in our database (see PR https://github.com/alphagov/forms-admin/pull/469) and to better express the intention of the configuration variable, I think we should stick with provider.

jamie-o-wilkinson commented on app/controllers/application_controller.rb:
```diff
@@ -20,51 +20,12 @@ class ApplicationController < ActionController::Base
     end
   end
 
-  def check_access
```
Didn't realise this was now only called from one place. Nice 👍

jamie-o-wilkinson commented on app/helpers/application_helper.rb:
```diff
@@ -65,11 +65,18 @@ def govuk_assets_path
   end
 
   def header_component_options(user:, can_manage_users:)
+    links = Hash.new({}).update( # rubocop:disable Rails/SaveBang
```
It took me a moment to figure out how this worked (though I may just be slow after not enough sleep last night 😪 )...might be worth a chat to see if there is a clearer way?

lfdebrux replied:
I would like a clearer way haha...

lfdebrux replied:
I've refactored to make more explicit, thanks for the nudge

lfdebrux commented on Gemfile:
```diff
@@ -14,6 +14,10 @@ gem "pg", "~> 1.5"
 
+# Used for handling authentication
+gem "gds-sso"
+gem "warden"
```
Yeah, gds-sso depends on Warden.

lfdebrux commented:
I've renamed things a bit, I decided that it would be better to be consistent in our configuration interface with what we save in our db, so I've changed the name of the configuration variable to `auth_provider`.

DavidBiddle reviewed (approved):
- Looks good overall 🎉 
- Makes application controller less complex 👏
- Adds integration tests 😻
